src/librustc_mir/interpret/memory.rs

```diff
@@ -785,10 +785,28 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
         assert_eq!(size.bytes() as usize as u64, size.bytes());
 
         let undef_mask = self.get(src.alloc_id)?.undef_mask.clone();
+        let get = |i| undef_mask.get(src.offset + Size::from_bytes(i));
         let dest_allocation = self.get_mut(dest.alloc_id)?;
 
+        // an optimization where we can just overwrite an entire range of definedness bits if
+        // they are going to be uniformly `1` or `0`.
+        if size.bytes() * repeat > UndefMask::BLOCK_SIZE {
+            let first = undef_mask.get(src.offset);
+            // check that all bits are the same as the first bit
+            // FIXME(oli-obk): consider making this a function on `UndefMask` and optimize it, too
+            if (1..size.bytes()).all(|i| get(i) == first) {
+                dest_allocation.undef_mask.set_range(
+                    dest.offset,
+                    dest.offset + size * repeat,
+                    first,
+                );
+                return Ok(())
```

Comment:
So if, say, the 999th bit is different, we'll just throw away all that time we spent and use the manual loop below?
Couldn't we instead make it determine the first *different* bit, then using the `set_range` for everything until that, and then copy the rest?

But also, it seems to me a proper `memcpy` should be faster than what you are trying here. Why can't we `memcpy` the undef mask?

Reply:
We can probably do a `memcpy` for `size`s that are powers of two (or where you can add up `size`s to reach a multiple of `64`. Otherwise copying gets very weird.

Reply:
Ah right, every bit of these `u64` corresponds to 1 byte... hm.

Something shift-based would work, I guess. Not sure if it's worth it. Just, I am not sure if this "all the same bit"-base optimization is worth it in terms of code complexity vs. the number of cases that'll actually see a speed-up.

What about doing it all in ranges? Like, iterate over all the bits, and every time the value changes, use the set function to set the range we juts saw? That'll at least also apply for cases like "<long initialized part> <uninitialized part> <long initialized part>".

Reply:
oh that part of the code is no good anyway. It's completely trashing any caches. We have an outer loop over the bits in the source value and then an inner loop over *all* the copies setting just the current bit from the source.

Even if we swap the loops, we'll want to do as little work as possible, and the range suggestion would still be setting bits one by one, so the effect would probably be rather minimal and possibly offset by the additional work done for counting the bits for every copy of the source value.

We can probably precompute the optimization by having an "offset until next undef inversion" vector. I'll try that, but the code complexity will go up further.

Reply:
> Even if we swap the loops, we'll want to do as little work as possible, and the range suggestion would still be setting bits one by one, so the effect would probably be rather minimal and possibly offset by the additional work done for counting the bits for every copy of the source value.

It wouldn't set the bits one-by-one if the range is larger enough to contain an entire `u64`. And, in fact, `set_range` could use bit mask operations to be much faster even for partial blocks.

Reply:
it will only do that if a single element's range is larger than 64 bytes. The common case is not types that are larger than 64 bytes.

but yes, there are many more aventues for optimizing setting partial bit ranges

Reply:
I did this, and the 24 second stage0 example is now a 10 second example on stage 1. Note that I think my stage1 is generally slower than stage0 because of extra debug stuff compiled in, so the actual gains should be even higher.